unix: add -X realtime option for macOS precision timers

dlech opened the pull request:
The reasoning for this has already been discussed in https://github.com/micropython/micropython/issues/8621.

There is also an additional commit to fix a conflicting type name that caused a compile error when `mach` headers were included.

dlech committed:
ports/mpthreadport: Replace thread_t with mp_thread_t.

This adds the `mp_` prefix to the `thread_t` type. The name `thread_t`
conflicts with the same in `mach/mach_types.h` on macOS.

Signed-off-by: David Lechner <david@lechnology.com>

dlech committed:
ports/unix: Implement -X realtime on macOS.

This adds a new command line option to the unix port `-X realtime` to
enable realtime priority on threads. This enables high precision timers
for applications that need more accurate timers.

Related docs: https://developer.apple.com/library/archive/technotes/tn2169/_index.html
Fixes: https://github.com/micropython/micropython/issues/8621

Signed-off-by: David Lechner <david@pybricks.com>

dpgeorge commented:
To test this, remove the uasyncio exclusions in `tools/ci.sh:ci_unix_macos_run_tests`

dlech commented:
> To test this, remove the uasyncio exclusions in `tools/ci.sh:ci_unix_macos_run_tests`

Added this to the last commit. Looks green on the first run.

dpgeorge commented on py/mpthread.h:
```diff
@@ -58,4 +60,10 @@ void mp_thread_mutex_unlock(mp_thread_mutex_t *mutex);
+#if defined(__APPLE__)
+extern bool mp_thread_is_realtime_enabled;
+void mp_thread_set_realtime(void);
+#endif
```
IMO this chunk should go in `ports/unix/mpthreadport.h`.

dlech replied:
done

dpgeorge commented on tests/run-tests.py:
```diff
@@ -165,6 +165,8 @@ def send_get(what):
             if args.heapsize is not None:
                 cmdlist.extend(["-X", "heapsize=" + args.heapsize])
+            if sys.platform == "darwin":
+                cmdlist.extend(["-X", "realtime"])
```
Please put this change, and the change to `tools/ci.sh`, in a separate commit.

dlech replied:
done

dlech committed:
tests: Enable -X realtime option for macOS tests.

This enables the new `-X realtime` runtime option when running tests
on macOS. This causes MicroPython to configure all threads to be high
priority so that they are allowed to use high precision timers. This
makes tests that depend on the passage of time more likely to succeed.

CI tests that were disabled because of this are now enabled again.

Signed-off-by: David Lechner <david@pybricks.com>
